modules/platform/nuxeo-platform-web-common/src/main/java/org/nuxeo/ecm/platform/web/common/requestcontroller/filter/IdempotentRequestHandler.java

```diff
@@ -0,0 +1,144 @@
+        // retrieve potential configuration from filter
+        String ttlConfig = filterConfig.getInitParameter(IdempotentRequestHandler.TTL_SECONDS_PROPERTY);
+        if (ttlConfig != null) {
+            try {
+                ttl = Integer.valueOf(ttlConfig);
+            } catch (NumberFormatException e) {
+                log.error(e, e);
+            }
+        }
+        String storeConfig = filterConfig.getInitParameter(IdempotentRequestHandler.STORE_PROPERTY);
+        if (storeConfig != null) {
+            store = storeConfig;
+        }
+        this.store = store;
+        this.ttl = ttl;
+    }
+
+    public IdempotentRequestHandler(String storename, int ttl) {
+        this.store = storename;
+        this.ttl = ttl;
+    }
+
+    public void handle(HttpServletRequest request, HttpServletResponse response, FilterChain chain)
+            throws IOException, ServletException {
+        if (!"POST".equals(request.getMethod())) {
+            chain.doFilter(request, response);
+            return;
+        }
+        String key = request.getHeader(IdempotentRequestHandler.HEADER_KEY);
+        if (key == null) {
+            chain.doFilter(request, response);
+            return;
+        }
+        KeyValueStore store = Framework.getService(KeyValueService.class).getKeyValueStore(this.store);
+        String storedContent = store.getString(IdempotentRequestHandler.CONTENT_PREFIX + key);
+        if (storedContent == null) {
+            // new request
+            boolean cleanup = false;
+            store.put(IdempotentRequestHandler.CONTENT_PREFIX + key, IdempotentRequestHandler.INPROGRESS_MARKER, ttl);
+            try {
+                CopyingResponseWrapper wrapper = new CopyingResponseWrapper(response);
+                chain.doFilter(request, wrapper);
+                store.put(IdempotentRequestHandler.CONTENT_PREFIX + key, wrapper.getCaptureAsString(), ttl);
+                store.put(IdempotentRequestHandler.STATUS_PREFIX + key, String.valueOf(wrapper.getStatus()), ttl);
+            } catch (IOException | ServletException e) {
+                cleanup = true;
+                throw e;
+            } finally {
+                if (cleanup || response.getStatus() >= 400) {
+                    // error request: cleanup store
+                    store.put(IdempotentRequestHandler.CONTENT_PREFIX + key, (String) null, ttl);
+                    store.put(IdempotentRequestHandler.STATUS_PREFIX + key, (String) null, ttl);
+                }
+            }
+        } else if (IdempotentRequestHandler.INPROGRESS_MARKER.equals(storedContent)) {
+            // request already in progress -> conflict
+            response.sendError(HttpServletResponse.SC_CONFLICT,
+                    String.format("Idempotent request already in progress for key '%s'", key));
+        } else {
+            // request already done: return stored result
+            response.getWriter().write(storedContent);
```

Comment:
As mentioned you should use bytes, not string. So use `getOutputStream` here instead of `getWriter`.